### ISSUE TITLE
Interface update tracker

### ISSUE THREAD

**Issue by shamblett** [shamblett/ethereum#4]
Raised for users to add what they would like the package to support, also this will track any re-implementation work if the Ethereum RPC interface is updated.

**Comment by gkd720** [comment]
I would like to see support for the "personal" subset of the "Management APIs" (issue #2). They are documented at [https://github.com/ethereum/go-ethereum/wiki/Management-APIs#personal](url). If the work is significant or excessive, I would request them in the following order: `sign`, `sendTransaction`, `ecRecover`, `unlockAccount`, `newAccount`, and `listAccounts`, though the remainder would be useful too. I would also be interested in your invocation in dart as I'm just coming up to speed in it, so I guess I can just look at your test file. Thanks for looking into this.

**Comment by shamblett** [comment]
OK, no problem with this, I'll start with the methods you listed and eventually add the whole of it. I can re-publish the package as I go so we don't have to wait for an all or nothing approach.

The API I use is Dart world, I expect Dart constructs to be passed, i.e. int, float, String, bool etc to the method call, if the method states that some parameters are mandatory these are checked before the RPC call. Similarly on return all info is converted back into Dart world, whether these be errors or not. I don't expect the users to be decoding their own JSON or poking around in byte buffers etc to see whats happened.

**Comment by shamblett** [comment]
The Personal methods of the Admin interface are now implemented, re-published the package at 3.0.0.
The API has changed a little, from now on the API calls are namespaced as per the geth rpcapi names, so the original calls are now preceded with .eth, the new admin ones with .admin. Previously you did 

await client.protocolVersion();

now you do 

await client.eth.protocolVersion();

This allows the code to not include all the calls in the client class which would just get longer and longer, this is better for maintainability and documentation. There is however a blowback on the user as their code must now be updated, I think for this change the pain is OK.

I'll carry on implementing the admin API's in no particular order, unless you have a priority list.

**Comment by gkd720** [comment]
Thanks for implementing these. Sorry it took so long to start to use them. I assigned a "client", and then I looked into using/calling personalSendTransaction. I had some trouble with the async/await/future world, but I think I see how to call it. Are there any other examples available? I was surprised as I thought about the arguments and return value. The basic etc.sendTransaction returns a transaction hash, which I need, but the admin one returns a boolean. Can a hash be extracted somehow? Also, the admin one only has from, to, and passphrase. Where do the rest of the transaction fields get assigned? Also, even in the eth one, I don't see the "input" field, which I use to store some string data in the transaction stored in the blockchain. Does that exist? Thanks for looking into this.

**Comment by shamblett** [comment]
Ah yes, looking again I need to update these methods to return the correct parameters, the management page is different from the main JSON RPC wiki page in that it doesn't have a returns section, it just seems to be implied under the call itself. I'll look at your other comments also.

**Comment by gkd720** [comment]
Great. Thanks. Yeah, I'm kind of translating a javascript version/flow I have into Flutter/Dart, and it was based on the "personal" set of methods in the Management APIs listed at [https://github.com/ethereum/go-ethereum/wiki/Management-APIs](url). My primary usage involves personal_sendTransaction, personal_sign, and personal_ecRecover. Transaction fields can be seen in a getTransaction example at [https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethgettransaction](url), where "input" is either contract code or just data, and is determined by whether the "to" address is null or not.

**Comment by shamblett** [comment]
I've updated personalSendTransaction from [here](https://wiki.parity.io/JSONRPC-personal-module#personal_sendtransaction), plese have a look at this, its now much closer to the eth version but with a passphrase, still can't see an 'input' field though.

**Comment by gkd720** [comment]
Thanks. I'll check it out.
I now see what I was thinking with the field names. The Ethereum Yellow Paper (p.3, sec 4.2) describes 2 different names for arbitrary data associated with sending a transaction. If there is no "to" address specified, the field is called "init", and is the contract code to be established. If there is a "to" address specified, the field is called "data", and is an arbitrary byte string. From the link above for a "getTransaction" call (which I now see was specified incorrectly, but can be copied/pasted), this init/data information is displayed as the "input" field, so I guess "input" was chosen as the more generic term for that field's value. In any documentation  for the JSON RPC API, a transaction's arbitrary data input is always called "data" at the JSON object level in the HTTP request, as you mentioned, with its interpretation determined by the existence of a "to" address, and also detailed [here](https://stackoverflow.com/questions/35076404/send-transactions-to-ethereum-contracts-via-json-rpc-api). So, "data" sounds good. I'll report back any progress. Thanks for working on this.

Update: 
I was poking around in your code, and I didn't see any changes. That was in the "master" branch, so I clicked on "development" and I see updates. Will these changes make it into "master" at some point? Which version do I get with a package update from redoing "flutter packages get":?

Also, looking at the code, I see the other fields now exist. So I don't build up a complete "transaction" map before calling it, but just add the parameter that is a map of these new fields?

**Comment by shamblett** [comment]
The other fields, bar one, are the same as the eth function so you can use it in the same way, you just supply whatever parameters you need, the method builds the map.

Yes, I'll merge development into master and re publish the package, I can do this now if your OK with this, even if we find something else later I can easily republish again etc.

**Comment by gkd720** [comment]
Sounds good. I'll start mapping out how to incorporate into my application. Please merge into "master" and republish. Thanks for working on this.

**Comment by shamblett** [comment]
Republished at 3.1.0, see how you get on, please update this issue with any probs, requests etc.

**Comment by gkd720** [comment]
I looked at your example and see how to create a "client". But I had trouble trying to access the admin methods. So I think I don't fully understand how to use this. Here's some code and results. What am I doing wrong? Thanks.

```
import 'package:ethereum/ethereum.dart';
import 'package:ethereum/ethereum_server_client.dart';
. . .
  // Set up IP/port of Ethereum node to talk to.
  final EthereumServerClient client = new EthereumServerClient.withConnectionParameters("localhost",8545);
  // Set up access to admin APIs
  final EthereumApiAdmin admin = new EthereumApiAdmin(client);
. . .

Compiler message:
lib/create_bottle_screen.dart:43:55: Error: Can't access 'this' in a field initializer to read 'client'.
  final EthereumApiAdmin admin = new EthereumApiAdmin(client);
                                                      ^^^^^^
```

**Comment by shamblett** [comment]
Your first line is correct, you then access the API methods through the client, example 

final String version = await client.eth.protocolVersion();

Note the 'eth' bit, this is a namespace for the eth API, there is also an admin API

 await client.admin.personalUnlockAccount(lockAddress, 'password');

There are examples of calling every method in the ethereum_common.dart file in the test directory

The actual API classes are not meant to be instantiated, I'll fix this so they can't be.

**Comment by gkd720** [comment]
Hmmm, still having trouble. I put protocolVersion example call line just below the client assignment, but I get an error of " A value of type 'Future<String>' can't be assigned to a variable of type 'String' ". Searching suggests I have to farm it out to a Future method:

```
class CreateBottleScreenState extends State<CreateBottleScreen> {
. . .
  final EthereumServerClient client = new EthereumServerClient.withConnectionParameters("localhost",8545);
. . .
  String protocolVersion;
. . .
  Widget build(BuildContext context) {
(now in a button)
                onPressed: () {
                    protocolVersionFuture();
. . .
  }     (end of build)
  Future protocolVersionFuture() async {
    String protocolVersion = await client.eth.protocolVersion();
    setState(() => this.protocolVersion = protocolVersion);
  }
}     (end of class)
```

I didn't infer this from your test code example. Did I miss something? Am I just doing things the hard way? 
Thanks.

**Comment by shamblett** [comment]
This code looks very fluttery, I don't use flutter but you could try annotating your types, as in 

Fututre protocolVersionFuture() 

becomes

Future<String> protocolVersionFuture()

Also assigning straight to the property may help i.e.

this.protocolVersion = await client.eth.protocolVersion();

**Comment by gkd720** [comment]
Yes, this is a Flutter application. My style of invocation above seems to be a simple way to go. The protocolVersion method invocation works using this style, so I tried client.admin.personalSendTransaction. But this gets a compile error in VS Code complaining about argument count, and I'm stuck. Details [here](https://stackoverflow.com/questions/53731646/how-to-upgrade-packages-in-vscode-flutter). Any thoughts? Thanks.

Update:

I tried calling anyway. It actually does occasionally actually compile!

Invocation:

```
                    print("Calling submitTxCreate()");
                    submitTxCreate();
                    print("Called submitTxCreate()");
```

Definition:

```
  Future submitTxCreate() async {
    print("Starting submitTxCreate");
    hashTxCreate = await client.admin.personalSendTransaction(BigInt.parse(currentAddress), currentPassword,{});
    print("hashTxCreate= " + hashTxCreate.toString());
    print("Ending submitTxCreate");
  }
```

Results:

```
I/flutter (24102): Calling submitTxCreate()
I/flutter (24102): Starting submitTxCreate
I/flutter (24102): Called submitTxCreate()
. . .
E/flutter (24102): [ERROR:flutter/shell/common/shell.cc(184)] Dart Error: Unhandled exception:
E/flutter (24102): NoSuchMethodError: Class 'EthereumApiAdmin' has no instance method 'personalSendTransaction' with matching arguments.
E/flutter (24102): Receiver: Instance of 'EthereumApiAdmin'
E/flutter (24102): Tried calling: personalSendTransaction(Instance of '_BigIntImpl', "prod1pw", _LinkedHashMap len:0)
E/flutter (24102): Found: personalSendTransaction(BigInt, String, {BigInt to, BigInt data, int gas, int gasPrice, int value, int nonce, int condition, bool conditionIsTimestamp}) => Future<BigInt>
E/flutter (24102): #0      Object.noSuchMethod (dart:core/runtime/libobject_patch.dart:50:5)
```

Is an empty transaction deadly? I don't see anywhere an empty transaction should get an error.

**Comment by gkd720** [comment]
Hmmm, not sure if you've seen my response under "Issues", but I've tried various things and I can't seem to fix it. It still looks like the VS Code compiler is using version 3.0.0, even though I require 3.1.0 in the pubspec (with/without leading caret). I even filled out the Map containing all keys. Any thoughts, guidance, or experiments I might try? Thanks.


Compiler message:
lib/create_bottle_screen.dart:209:69: Error: Too many positional arguments: 2 allowed, but 3 found.
Try removing the extra positional arguments.
BigInt hashTxCreate = await client.admin.personalSendTransaction(BigInt.parse(currentAddress), currentPassword,{'to':BigInt.from(1),'data':BigInt.from(2),'gas':3,'gasPrice':4,'value':1,'nonce':5,'condition':6,'cconditionIsTimestamp':false});
^     (the caret is really under the left parenthesis before the first argument)
Compiler failed on /Users/gkd/Programs/wine_track/lib/main.dart

**Comment by shamblett** [comment]
In the call above you have 'cconditionIsTimestamp' rather than 'conditionIsTimestamp', looks as though this is causing the error although admittedly the error text could be better.

**Comment by gkd720** [comment]
Hmmm, I would have thought that would not actually cause an error since it would have just supplied a key that is not needed. But I don't think it's even getting that far since it balks at the method call itself. Anyway, I fixed this, but the error is the same. 

Wondering what you would see on your end, I looked at the test code. It looks like your test invocation still calls it with just 2 parameters (address, password). In the functional code, it's specified with 3 parameters, and the last (remaining transaction fields) doesn't seem optional. Shouldn't your testing have seen a problem?

It still seems like I'm operating with a mix of 3.0.0 code and 3.1.0 code, since when I call that method with just 2 arguments, it works.  And hovering over the invocation with the mouse in my source code shows its definition, with 3 arguments! Are there any package files, besides pubspec.yaml and pubspec.lock, that specify the version of the package to be used? Should I just try removing the 3.0.0 files from the pub cache?

Any other thoughts? Thanks for looking into this.

Update: 
OK, never mind. This was my mistake. I interpreted the declaration's "{ type:variablename, . . .}" as a map/hash. In reality, they're Dart's optional named parameters. Just using any of them without the wrapping braces compiles clean.
